// @tarojs/cli/dist/util/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs = require("fs-extra");
const path = require("path");
const stream_1 = require("stream");
const crypto = require("crypto");
const os = require("os");
const child_process = require("child_process");
const chalk_1 = require("chalk");
const lodash_1 = require("lodash");
const minimatch = require("minimatch");
const t = require("babel-types");
const yauzl = require("yauzl");
const findWorkspaceRoot = require("find-yarn-workspace-root");
const chokidar = require("chokidar");
const babel_1 = require("../config/babel");
const uglify_1 = require("../config/uglify");
const constants_1 = require("./constants");
const npm_1 = require("./npm");
const execSync = child_process.execSync;
function isNpmPkg(name) {
    if (/^(\.|\/)/.test(name)) {
        return false;
    }
    return true;
}
exports.isNpmPkg = isNpmPkg;
function isAliasPath(name, pathAlias = {}) {
    const prefixs = Object.keys(pathAlias);
    if (prefixs.length === 0) {
        return false;
    }
    return prefixs.includes(name) || (new RegExp(`^(${prefixs.join('|')})/`).test(name));
}
exports.isAliasPath = isAliasPath;
function replaceAliasPath(filePath, name, pathAlias = {}) {
    // åŽç»­çš„ path.join åœ¨é‡åˆ°ç¬¦å·é“¾æŽ¥æ—¶å°†ä¼šè§£æžä¸ºçœŸå®žè·¯å¾„ï¼Œå¦‚æžœ
    // è¿™é‡Œçš„ filePath æ²¡æœ‰åšåŒæ ·çš„å¤„ç†ï¼Œå¯èƒ½ä¼šå¯¼è‡´ import æŒ‡å‘
    // æºä»£ç æ–‡ä»¶ï¼Œå¯¼è‡´æ–‡ä»¶è¢«æ„å¤–ä¿®æ”¹
    filePath = fs.realpathSync(filePath);
    const prefixs = Object.keys(pathAlias);
    if (prefixs.includes(name)) {
        return promoteRelativePath(path.relative(filePath, fs.realpathSync(resolveScriptPath(pathAlias[name]))));
    }
    const reg = new RegExp(`^(${prefixs.join('|')})/(.*)`);
    name = name.replace(reg, function (m, $1, $2) {
        return promoteRelativePath(path.relative(filePath, path.join(pathAlias[$1], $2)));
    });
    return name;
}
exports.replaceAliasPath = replaceAliasPath;
function promoteRelativePath(fPath) {
    const fPathArr = fPath.split(path.sep);
    let dotCount = 0;
    fPathArr.forEach(item => {
        if (item.indexOf('..') >= 0) {
            dotCount++;
        }
    });
    if (dotCount === 1) {
        fPathArr.splice(0, 1, '.');
        return fPathArr.join('/');
    }
    if (dotCount > 1) {
        fPathArr.splice(0, 1);
        return fPathArr.join('/');
    }
    return fPath.replace(/\\/g, '/');
}
exports.promoteRelativePath = promoteRelativePath;
exports.homedir = os.homedir;
function getRootPath() {
    return path.resolve(__dirname, '../../');
}
exports.getRootPath = getRootPath;
function getTaroPath() {
    const taroPath = path.join(exports.homedir(), '.taro');
    if (!fs.existsSync(taroPath)) {
        fs.ensureDirSync(taroPath);
    }
    return taroPath;
}
exports.getTaroPath = getTaroPath;
function getConfig() {
    const configPath = path.join(getTaroPath(), 'config.json');
    if (fs.existsSync(configPath)) {
        return require(configPath);
    }
    return {};
}
exports.getConfig = getConfig;
function getSystemUsername() {
    const userHome = exports.homedir();
    const systemUsername = process.env.USER || path.basename(userHome);
    return systemUsername;
}
exports.getSystemUsername = getSystemUsername;
function getPkgVersion() {
    return require(path.join(getRootPath(), 'package.json')).version;
}
exports.getPkgVersion = getPkgVersion;
function getPkgItemByKey(key) {
    const packageMap = require(path.join(getRootPath(), 'package.json'));
    if (Object.keys(packageMap).indexOf(key) === -1) {
        return {};
    }
    else {
        return packageMap[key];
    }
}
exports.getPkgItemByKey = getPkgItemByKey;
function printPkgVersion() {
    const taroVersion = getPkgVersion();
    console.log(`ðŸ‘½ Taro v${taroVersion}`);
    console.log();
}
exports.printPkgVersion = printPkgVersion;
function shouldUseYarn() {
    try {
        execSync('yarn --version', { stdio: 'ignore' });
        return true;
    }
    catch (e) {
        return false;
    }
}
exports.shouldUseYarn = shouldUseYarn;
function shouldUseCnpm() {
    try {
        execSync('cnpm --version', { stdio: 'ignore' });
        return true;
    }
    catch (e) {
        return false;
    }
}
exports.shouldUseCnpm = shouldUseCnpm;
function isEmptyObject(obj) {
    if (obj == null) {
        return true;
    }
    for (const key in obj) {
        if (obj.hasOwnProperty(key)) {
            return false;
        }
    }
    return true;
}
exports.isEmptyObject = isEmptyObject;
function urlJoin(...agrs) {
    function normalize(str) {
        return str
            .replace(/([/]+)/g, '/')
            .replace(/\/\?(?!\?)/g, '?')
            .replace(/\/#/g, '#')
            .replace(/:\//g, '://');
    }
    const joined = [].slice.call(agrs, 0).join('/');
    return normalize(joined);
}
exports.urlJoin = urlJoin;
function resolveScriptPath(p) {
    const realPath = p;
    const taroEnv = process.env.TARO_ENV;
    const SCRIPT_EXT = constants_1.JS_EXT.concat(constants_1.TS_EXT);
    for (let i = 0; i < SCRIPT_EXT.length; i++) {
        const item = SCRIPT_EXT[i];
        if (taroEnv) {
            if (fs.existsSync(`${p}.${taroEnv}${item}`)) {
                return `${p}.${taroEnv}${item}`;
            }
            if (fs.existsSync(`${p}${path.sep}index.${taroEnv}${item}`)) {
                return `${p}${path.sep}index.${taroEnv}${item}`;
            }
            if (fs.existsSync(`${p.replace(/\/index$/, `.${taroEnv}/index`)}${item}`)) {
                return `${p.replace(/\/index$/, `.${taroEnv}/index`)}${item}`;
            }
        }
        if (fs.existsSync(`${p}${item}`)) {
            return `${p}${item}`;
        }
        if (fs.existsSync(`${p}${path.sep}index${item}`)) {
            return `${p}${path.sep}index${item}`;
        }
    }
    return realPath;
}
exports.resolveScriptPath = resolveScriptPath;
function resolvePureScriptPath(p) {
    const realPath = p;
    const SCRIPT_EXT = constants_1.JS_EXT.concat(constants_1.TS_EXT);
    for (let i = 0; i < SCRIPT_EXT.length; i++) {
        const item = SCRIPT_EXT[i];
        if (fs.existsSync(`${p}${item}`)) {
            return `${p}${item}`;
        }
        if (fs.existsSync(`${p}${path.sep}index${item}`)) {
            return `${p}${path.sep}index${item}`;
        }
    }
    return realPath;
}
exports.resolvePureScriptPath = resolvePureScriptPath;
function resolveQuickappFilePath(p) {
    for (let i = 0; i < constants_1.UX_EXT.length; i++) {
        const item = constants_1.UX_EXT[i];
        if (fs.existsSync(`${p}${item}`)) {
            return `${p}${item}`;
        }
        if (fs.existsSync(`${p}${path.sep}index${item}`)) {
            return `${p}${path.sep}index${item}`;
        }
    }
    return p;
}
exports.resolveQuickappFilePath = resolveQuickappFilePath;
function processUxContent(contents, cb) {
    const reg = /(<script(?:(?=\s)[\s\S]*?["'\s\w\/\-]>|>))([\s\S]*?)(?=<\/script\s*>|$)|(<style(?:(?=\s)[\s\S]*?["'\s\w\/\-]>|>))([\s\S]*?)(?=<\/style\s*>|$)|<(image)\s+[\s\S]*?["'\s\w\/\-](?:>|$)|(<import(?:(?=\s)[\s\S]*?["'\s\w\/\-]>|>))([\s\S]*?)(?=<\/import\s*>|$)/ig;
    contents = contents.replace(reg, function (m, $1, $2, $3, $4, $5, $6, $7, $8, $9, $10) {
        if ($1) {
            $1 = $1.replace(/(\ssrc\s*=\s*)('[^']+'|"[^"]+"|[^\s\/>]+)/ig, function (m, prefix, value) {
                if (typeof cb === 'function') {
                    value = cb(value);
                }
                return prefix + value;
            });
            m = $1 + $2;
        }
        else if ($3) {
            $3 = $3.replace(/(\ssrc\s*=\s*)('[^']+'|"[^"]+"|[^\s\/>]+)/ig, function (m, prefix, value) {
                if (typeof cb === 'function') {
                    value = cb(value);
                }
                return prefix + value;
            });
            m = $3 + $4;
        }
        else if ($5) {
            m = m.replace(/(src\s*=\s*)('[^']+'|"[^"]+"|[^\s\/>]+)/ig, function (m, prefix, value) {
                if (typeof cb === 'function') {
                    value = cb(value);
                }
                return prefix + value;
            });
        }
        else if ($6) {
            $6 = $6.replace(/(\ssrc\s*=\s*)('[^']+'|"[^"]+"|[^\s\/>]+)/ig, function (m, prefix, value) {
                if (typeof cb === 'function') {
                    value = cb(value);
                }
                return prefix + value;
            });
            m = $6 + $7;
        }
        return m;
    });
    return contents;
}
exports.processUxContent = processUxContent;
function resolveStylePath(p) {
    const realPath = p;
    const removeExtPath = p.replace(path.extname(p), '');
    const taroEnv = process.env.TARO_ENV;
    for (let i = 0; i < constants_1.CSS_EXT.length; i++) {
        const item = constants_1.CSS_EXT[i];
        if (taroEnv) {
            if (fs.existsSync(`${removeExtPath}.${taroEnv}${item}`)) {
                return `${removeExtPath}.${taroEnv}${item}`;
            }
        }
        if (fs.existsSync(`${p}${item}`)) {
            return `${p}${item}`;
        }
    }
    return realPath;
}
exports.resolveStylePath = resolveStylePath;
function isDifferentArray(a, b) {
    if (!Array.isArray(a) || !Array.isArray(b)) {
        return true;
    }
    if (a.length !== b.length) {
        return true;
    }
    a = a.sort();
    b = b.sort();
    for (let i = 0; i < a.length; i++) {
        if (a[i] !== b[i]) {
            return true;
        }
    }
    return false;
}
exports.isDifferentArray = isDifferentArray;
function checksum(buf, length) {
    if (!Buffer.isBuffer(buf)) {
        buf = Buffer.from(buf);
    }
    return crypto.createHash('md5').update(buf).digest('hex').slice(0, length || 8);
}
exports.checksum = checksum;
function printLog(type, tag, filePath) {
    const typeShow = constants_1.processTypeMap[type];
    const tagLen = tag.replace(/[\u0391-\uFFE5]/g, 'aa').length;
    const tagFormatLen = 8;
    if (tagLen < tagFormatLen) {
        const rightPadding = new Array(tagFormatLen - tagLen + 1).join(' ');
        tag += rightPadding;
    }
    const padding = '';
    filePath = filePath || '';
    if (typeof typeShow.color === 'string') {
        console.log(chalk_1.default[typeShow.color](typeShow.name), padding, tag, padding, filePath);
    }
    else {
        console.log(typeShow.color(typeShow.name), padding, tag, padding, filePath);
    }
}
exports.printLog = printLog;
function replaceContentEnv(content, env) {
    if (env && !isEmptyObject(env)) {
        for (const key in env) {
            const reg = new RegExp(`process.env.${key}`, 'g');
            content = content.replace(reg, env[key]);
        }
        return content;
    }
    return content;
}
exports.replaceContentEnv = replaceContentEnv;
function generateEnvList(env) {
    const res = {};
    if (env && !isEmptyObject(env)) {
        for (const key in env) {
            try {
                res[`process.env.${key}`] = JSON.parse(env[key]);
            }
            catch (err) {
                res[`process.env.${key}`] = env[key];
            }
        }
    }
    return res;
}
exports.generateEnvList = generateEnvList;
function replaceContentConstants(content, constants) {
    if (constants && !isEmptyObject(constants)) {
        for (const key in constants) {
            const reg = new RegExp(key, 'g');
            content = content.replace(reg, constants[key]);
        }
        return content;
    }
    return content;
}
exports.replaceContentConstants = replaceContentConstants;
function generateConstantsList(constants) {
    const res = {};
    if (constants && !isEmptyObject(constants)) {
        for (const key in constants) {
            if (lodash_1.isPlainObject(constants[key])) {
                res[key] = generateConstantsList(constants[key]);
            }
            else {
                try {
                    res[key] = JSON.parse(constants[key]);
                }
                catch (err) {
                    res[key] = constants[key];
                }
            }
        }
    }
    return res;
}
exports.generateConstantsList = generateConstantsList;
function cssImports(content) {
    let match;
    const results = [];
    content = String(content).replace(/\/\*.+?\*\/|\/\/.*(?=[\n\r])/g, '');
    while ((match = constants_1.CSS_IMPORT_REG.exec(content))) {
        results.push(match[2]);
    }
    return results;
}
exports.cssImports = cssImports;
function processStyleImports(content, adapter, processFn) {
    const style = [];
    const imports = [];
    const styleReg = new RegExp(`\\${constants_1.MINI_APP_FILES[adapter].STYLE}`);
    content = content.replace(constants_1.CSS_IMPORT_REG, (m, $1, $2) => {
        if (styleReg.test($2)) {
            style.push(m);
            imports.push($2);
            if (processFn) {
                return processFn(m, $2);
            }
            return '';
        }
        if (processFn) {
            return processFn(m, $2);
        }
        return m;
    });
    return {
        content,
        style,
        imports
    };
}
exports.processStyleImports = processStyleImports;
/*eslint-disable*/
const retries = (process.platform === 'win32') ? 100 : 1;
function emptyDirectory(dirPath, opts = { excludes: [] }) {
    if (fs.existsSync(dirPath)) {
        fs.readdirSync(dirPath).forEach(file => {
            const curPath = path.join(dirPath, file);
            if (fs.lstatSync(curPath).isDirectory()) {
                let removed = false;
                let i = 0; // retry counter
                do {
                    try {
                        if (!opts.excludes.length || !opts.excludes.some(item => curPath.indexOf(item) >= 0)) {
                            emptyDirectory(curPath);
                            fs.rmdirSync(curPath);
                        }
                        removed = true;
                    }
                    catch (e) {
                    }
                    finally {
                        if (++i < retries) {
                            continue;
                        }
                    }
                } while (!removed);
            }
            else {
                fs.unlinkSync(curPath);
            }
        });
    }
}
exports.emptyDirectory = emptyDirectory;
/* eslint-enable */
function recursiveFindNodeModules(filePath) {
    const dirname = path.dirname(filePath);
    const workspaceRoot = findWorkspaceRoot(dirname);
    const nodeModules = path.join(workspaceRoot || dirname, 'node_modules');
    if (fs.existsSync(nodeModules)) {
        return nodeModules;
    }
    return recursiveFindNodeModules(dirname);
}
exports.recursiveFindNodeModules = recursiveFindNodeModules;
exports.pascalCase = (str) => str.charAt(0).toUpperCase() + lodash_1.camelCase(str.substr(1));
function getInstalledNpmPkgPath(pkgName, basedir) {
    const resolvePath = require('resolve');
    try {
        return resolvePath.sync(`${pkgName}/package.json`, { basedir });
    }
    catch (err) {
        return null;
    }
}
exports.getInstalledNpmPkgPath = getInstalledNpmPkgPath;
function checkCliAndFrameworkVersion(appPath, buildAdapter) {
    return __awaiter(this, void 0, void 0, function* () {
        const pkgVersion = getPkgVersion();
        const frameworkName = `@tarojs/taro-${buildAdapter}`;
        const nodeModulesPath = recursiveFindNodeModules(path.join(appPath, constants_1.NODE_MODULES));
        const frameworkVersion = getInstalledNpmPkgVersion(frameworkName, nodeModulesPath);
        if (frameworkVersion) {
            if (frameworkVersion !== pkgVersion) {
                const taroCliPath = path.join(getRootPath(), 'package.json');
                const frameworkPath = path.join(nodeModulesPath, frameworkName, 'package.json');
                printLog("error" /* ERROR */, 'ç‰ˆæœ¬é—®é¢˜', `Taro CLI ä¸Žæœ¬åœ°å®‰è£…è¿è¡Œæ—¶æ¡†æž¶ ${frameworkName} ç‰ˆæœ¬ä¸ä¸€è‡´, è¯·ç¡®ä¿ç‰ˆæœ¬ä¸€è‡´ï¼`);
                printLog("remind" /* REMIND */, 'å‡çº§å‘½ä»¤', `å‡çº§åˆ°æœ€æ–°CLIï¼štaro update self   å‡çº§åˆ°æœ€æ–°ä¾èµ–åº“ï¼štaro update project`);
                printLog("remind" /* REMIND */, 'å‡çº§æ–‡æ¡£', `è¯·å‚è€ƒ "å¸¸ç”¨ CLI å‘½ä»¤"ä¸­"æ›´æ–°" ç« èŠ‚ï¼šhttps://taro-docs.jd.com/taro/docs/GETTING-STARTED.html`);
                console.log(``);
                console.log(`Taro CLIï¼š${getPkgVersion()}             è·¯å¾„ï¼š${taroCliPath}`);
                console.log(`${frameworkName}ï¼š${frameworkVersion}   è·¯å¾„ï¼š${frameworkPath}`);
                console.log(``);
                process.exit(1);
            }
        }
        else {
            printLog("warning" /* WARNING */, 'ä¾èµ–å®‰è£…', chalk_1.default.red(`é¡¹ç›®ä¾èµ– ${frameworkName} æœªå®‰è£…ï¼Œæˆ–å®‰è£…æœ‰è¯¯ï¼Œè¯·é‡æ–°å®‰è£…æ­¤ä¾èµ–ï¼`));
            process.exit(1);
        }
    });
}
exports.checkCliAndFrameworkVersion = checkCliAndFrameworkVersion;
function getInstalledNpmPkgVersion(pkgName, basedir) {
    const pkgPath = getInstalledNpmPkgPath(pkgName, basedir);
    if (!pkgPath) {
        return null;
    }
    return fs.readJSONSync(pkgPath).version;
}
exports.getInstalledNpmPkgVersion = getInstalledNpmPkgVersion;
function traverseObjectNode(node, buildAdapter, parentKey) {
    if (node.type === 'ClassProperty' || node.type === 'ObjectProperty') {
        const properties = node.value.properties;
        const obj = {};
        properties.forEach(p => {
            let key = t.isIdentifier(p.key) ? p.key.name : p.key.value;
            if (constants_1.CONFIG_MAP[buildAdapter][key] === false) {
                return;
            }
            if (parentKey !== 'usingComponents' && constants_1.CONFIG_MAP[buildAdapter][key]) {
                key = constants_1.CONFIG_MAP[buildAdapter][key];
            }
            obj[key] = traverseObjectNode(p.value, buildAdapter, key);
        });
        return obj;
    }
    if (node.type === 'ObjectExpression') {
        const properties = node.properties;
        const obj = {};
        properties.forEach(p => {
            let key = t.isIdentifier(p.key) ? p.key.name : p.key.value;
            if (constants_1.CONFIG_MAP[buildAdapter][key] === false) {
                return;
            }
            if (parentKey !== 'usingComponents' && constants_1.CONFIG_MAP[buildAdapter][key]) {
                key = constants_1.CONFIG_MAP[buildAdapter][key];
            }
            obj[key] = traverseObjectNode(p.value, buildAdapter, key);
        });
        return obj;
    }
    if (node.type === 'ArrayExpression') {
        return node.elements.map(item => traverseObjectNode(item, buildAdapter));
    }
    if (node.type === 'NullLiteral') {
        return null;
    }
    return node.value;
}
exports.traverseObjectNode = traverseObjectNode;
function copyFileSync(from, to, options) {
    const filename = path.basename(from);
    if (fs.statSync(from).isFile() && !path.extname(to)) {
        fs.ensureDirSync(to);
        if (from === path.join(to, filename)) {
            return;
        }
        return fs.copySync(from, path.join(to, filename), options);
    }
    if (from === to) {
        return;
    }
    fs.ensureDirSync(path.dirname(to));
    return fs.copySync(from, to, options);
}
exports.copyFileSync = copyFileSync;
function copyFiles(appPath, copyConfig) {
    copyConfig = copyConfig || { patterns: [], options: {} };
    if (copyConfig.patterns && copyConfig.patterns.length) {
        copyConfig.options = copyConfig.options || {};
        const globalIgnore = copyConfig.options.ignore;
        const projectDir = appPath;
        copyConfig.patterns.forEach(pattern => {
            if (pattern.from && pattern.to) {
                const from = path.join(projectDir, pattern.from);
                const to = path.join(projectDir, pattern.to);
                let ignore = pattern.ignore || globalIgnore;
                if (fs.existsSync(from)) {
                    const copyOptions = {};
                    if (ignore) {
                        ignore = Array.isArray(ignore) ? ignore : [ignore];
                        copyOptions.filter = src => {
                            let isMatch = false;
                            ignore && ignore.forEach(iPa => {
                                if (minimatch(path.basename(src), iPa)) {
                                    isMatch = true;
                                }
                            });
                            return !isMatch;
                        };
                    }
                    copyFileSync(from, to, copyOptions);
                    if (pattern.watch) {
                        const watcher = chokidar.watch(from, {
                            persistent: true,
                            ignoreInitial: true
                        });
                        watcher.on('change', (res) => {
                            copyFileSync(from, to, copyOptions);
                        });
                    }
                }
                else {
                    printLog("error" /* ERROR */, 'æ‹·è´å¤±è´¥', `${pattern.from} æ–‡ä»¶ä¸å­˜åœ¨ï¼`);
                }
            }
        });
    }
}
exports.copyFiles = copyFiles;
function isQuickappPkg(name, quickappPkgs = []) {
    const isQuickappPkg = /^@(system|service)\.[a-zA-Z]{1,}/.test(name);
    let hasSetInManifest = false;
    quickappPkgs.forEach(item => {
        if (item.name === name.replace(/^@/, '')) {
            hasSetInManifest = true;
        }
    });
    if (isQuickappPkg && !hasSetInManifest) {
        printLog("error" /* ERROR */, 'å¿«åº”ç”¨', `éœ€è¦åœ¨ ${chalk_1.default.bold('project.quickapp.json')} æ–‡ä»¶çš„ ${chalk_1.default.bold('features')} é…ç½®ä¸­æ·»åŠ  ${chalk_1.default.bold(name)}`);
    }
    return isQuickappPkg;
}
exports.isQuickappPkg = isQuickappPkg;
function generateQuickAppUx({ script, template, style, imports }) {
    let uxTxt = '';
    if (imports && imports.size) {
        imports.forEach(item => {
            uxTxt += `<import src='${item.path}' name='${item.name}'></import>\n`;
        });
    }
    if (style) {
        if (constants_1.REG_STYLE.test(style)) {
            uxTxt += `<style src="${style}"></style>\n`;
        }
        else {
            uxTxt += `<style>\n${style}\n</style>\n`;
        }
    }
    if (template) {
        uxTxt += `<template>\n${template}\n</template>\n`;
    }
    if (script) {
        uxTxt += `<script>\n${script}\n</script>\n`;
    }
    return uxTxt;
}
exports.generateQuickAppUx = generateQuickAppUx;
exports.recursiveMerge = (src, ...args) => {
    return lodash_1.mergeWith(src, ...args, (value, srcValue, key, obj, source) => {
        const typeValue = typeof value;
        const typeSrcValue = typeof srcValue;
        if (typeValue !== typeSrcValue)
            return;
        if (Array.isArray(value) && Array.isArray(srcValue)) {
            return value.concat(srcValue);
        }
        if (typeValue === 'object') {
            return exports.recursiveMerge(value, srcValue);
        }
    });
};
exports.mergeVisitors = (src, ...args) => {
    const validFuncs = ['exit', 'enter'];
    return lodash_1.mergeWith(src, ...args, (value, srcValue, key, object, srcObject) => {
        if (!object.hasOwnProperty(key) || !srcObject.hasOwnProperty(key)) {
            return undefined;
        }
        const shouldMergeToArray = validFuncs.indexOf(key) > -1;
        if (shouldMergeToArray) {
            return lodash_1.flatMap([value, srcValue]);
        }
        const [newValue, newSrcValue] = [value, srcValue].map(v => {
            if (typeof v === 'function') {
                return {
                    enter: v
                };
            }
            else {
                return v;
            }
        });
        return exports.mergeVisitors(newValue, newSrcValue);
    });
};
exports.applyArrayedVisitors = obj => {
    let key;
    for (key in obj) {
        const funcs = obj[key];
        if (Array.isArray(funcs)) {
            obj[key] = (astPath, ...args) => {
                funcs.forEach(func => {
                    func(astPath, ...args);
                });
            };
        }
        else if (typeof funcs === 'object') {
            exports.applyArrayedVisitors(funcs);
        }
    }
    return obj;
};
function unzip(zipPath) {
    return new Promise((resolve, reject) => {
        yauzl.open(zipPath, { lazyEntries: true }, (err, zipfile) => {
            if (err)
                throw err;
            zipfile.on('close', () => {
                fs.removeSync(zipPath);
                resolve();
            });
            zipfile.readEntry();
            zipfile.on('error', (err) => {
                reject(err);
            });
            zipfile.on('entry', entry => {
                if (/\/$/.test(entry.fileName)) {
                    const fileNameArr = entry.fileName.replace(/\\/g, '/').split('/');
                    fileNameArr.shift();
                    const fileName = fileNameArr.join('/');
                    fs.ensureDirSync(path.join(path.dirname(zipPath), fileName));
                    zipfile.readEntry();
                }
                else {
                    zipfile.openReadStream(entry, (err, readStream) => {
                        if (err)
                            throw err;
                        const filter = new stream_1.Transform();
                        filter._transform = function (chunk, encoding, cb) {
                            cb(undefined, chunk);
                        };
                        filter._flush = function (cb) {
                            cb();
                            zipfile.readEntry();
                        };
                        const fileNameArr = entry.fileName.replace(/\\/g, '/').split('/');
                        fileNameArr.shift();
                        const fileName = fileNameArr.join('/');
                        const writeStream = fs.createWriteStream(path.join(path.dirname(zipPath), fileName));
                        writeStream.on('close', () => { });
                        readStream
                            .pipe(filter)
                            .pipe(writeStream);
                    });
                }
            });
        });
    });
}
exports.unzip = unzip;
let babelConfig;
function getBabelConfig(babel) {
    if (!babelConfig) {
        babelConfig = lodash_1.mergeWith({}, babel_1.default, babel, (objValue, srcValue) => {
            if (Array.isArray(objValue)) {
                return Array.from(new Set(srcValue.concat(objValue)));
            }
        });
    }
    return babelConfig;
}
exports.getBabelConfig = getBabelConfig;
function uglifyJS(resCode, filePath, root, uglify) {
    const uglifyPluginConfig = uglify || { enable: true };
    if (uglifyPluginConfig.enable) {
        const uglifyConfig = Object.assign(uglify_1.default, uglifyPluginConfig.config || {});
        const uglifyResult = npm_1.callPluginSync('uglifyjs', resCode, filePath, uglifyConfig, root);
        if (uglifyResult.error) {
            printLog("error" /* ERROR */, 'åŽ‹ç¼©é”™è¯¯', `æ–‡ä»¶${filePath}`);
            console.log(uglifyResult.error);
            return resCode;
        }
        return uglifyResult.code;
    }
    return resCode;
}
exports.uglifyJS = uglifyJS;
exports.getAllFilesInFloder = (floder, filter = []) => __awaiter(this, void 0, void 0, function* () {
    let files = [];
    const list = readDirWithFileTypes(floder);
    yield Promise.all(list.map((item) => __awaiter(this, void 0, void 0, function* () {
        const itemPath = path.join(floder, item.name);
        if (item.isDirectory) {
            const _files = yield exports.getAllFilesInFloder(itemPath, filter);
            files = [...files, ..._files];
        }
        else if (item.isFile) {
            if (!filter.find(rule => rule === item.name))
                files.push(itemPath);
        }
    })));
    return files;
});
function getUserHomeDir() {
    function homedir() {
        const env = process.env;
        const home = env.HOME;
        const user = env.LOGNAME || env.USER || env.LNAME || env.USERNAME;
        if (process.platform === 'win32') {
            return env.USERPROFILE || '' + env.HOMEDRIVE + env.HOMEPATH || home || '';
        }
        if (process.platform === 'darwin') {
            return home || (user ? '/Users/' + user : '');
        }
        if (process.platform === 'linux') {
            return home || (process.getuid() === 0 ? '/root' : (user ? '/home/' + user : ''));
        }
        return home || '';
    }
    return typeof os.homedir === 'function' ? os.homedir() : homedir();
}
exports.getUserHomeDir = getUserHomeDir;
function getTemplateSourceType(url) {
    if (/^github:/.test(url) || /^gitlab:/.test(url) || /^direct:/.test(url)) {
        return 'git';
    }
    else {
        return 'url';
    }
}
exports.getTemplateSourceType = getTemplateSourceType;
function readDirWithFileTypes(floder) {
    const list = fs.readdirSync(floder);
    const res = list.map(name => {
        const stat = fs.statSync(path.join(floder, name));
        return {
            name,
            isDirectory: stat.isDirectory(),
            isFile: stat.isFile()
        };
    });
    return res;
}
exports.readDirWithFileTypes = readDirWithFileTypes;
function extnameExpRegOf(filePath) {
    return new RegExp(`${path.extname(filePath)}$`);
}
exports.extnameExpRegOf = extnameExpRegOf;
function generateAlipayPath(filePath) {
    return filePath.replace(/@/g, '_');
}
exports.generateAlipayPath = generateAlipayPath;
